### ISSUE TITLE
Refactor: Extract Article rendering 

### ISSUE THREAD

**Issue by jaw6** [forem/forem#18649]
Currently, `Article`, `Comment`, and `DisplayAd` each have similar-but-subtly different rendering procedures. We'd like to unify these into one consistent, site-wide process. The first step will be to refactor, extracting `Article`'s current markdown-to-html code into a service object.

Release steps:
[ ] Thorough manual QA to ensure that nothing is broken on a Canary instance
[ ] Community Team / Product Heads up
[ ] Post release monitoring

**Comment by mirie** [comment]
@jaw6 I want to make sure we don't miss the manual QA piece -- how do you want to coordinate on this

**Comment by mirie** [comment]
@cc @maestromac / @fdocr for visibility re: post manual QA that we'll want to monitor closely once we deploy

**Comment by jaw6** [comment]
Good point! I suspect we ought to put the refactor behind a `FeatureFlag`. That way we can turn it on on a Canary and QA it there, without holding it up indefinitely.

**Comment by mirie** [comment]
@michael-tharrington (cc @benhalpern and @pkfrank): Just a heads up that we are close to rolling this out (**post manual QA**). The issue is rather technical but the tldr; is we are in the process of updating the rendering of Articles, Comments, and Display Ads so that they have on consistent process versus 3 different ones. 

This issue addresses step 1: Articles. Comments and DisplayAds (Billboards) will follow afterwards. 

As this is a change affecting Articles, we want to give you a heads up so that we can be aware of any support tickets incoming regarding Article creation, preview, display. We'll communicate over Slack in #community and #eng-headsup to ensure everyone is on the same page from Community Team <> Eng <> Platforms when we actually enable the change on production but hoping to get this live on production this week.

**Comment by mirie** [comment]
@jaw6 I'm not sure if you've been coordinating on Canary 1 QA deployment -- please let me know when it's available to do QA there so I can help out

**Comment by maestromac** [comment]
This is deployed to canary1 for QA

**Comment by mirie** [comment]
@jaw6 Let me know if it's easier for you if I record a Loom of my findings or if you have any questions about my findings so far.

**Comment by mirie** [comment]
@anujbhavsar96 I would love to have help [QA](https://docs.google.com/spreadsheets/d/1ypY8lLeIu3AiLIz741U1i4pkb_FCxqXCqZaP5VtTNwc/edit#gid=1302175926) on Canary1 if you have the bandwidth:

- Testing Create/Update/Delete functionality as applicable on Mobile app
- Do you think there's a need to double check published articles at mobile breakpoint?

**Comment by anujbhavsar96** [comment]
@mirie I can definitely help you with that ✨

Do you want me to test on Android and iOS app only? What about mobile browser?

> Do you think there's a need to double check published articles at mobile breakpoint?

I feel it's good to double-check. It won't take much time.

**Comment by jaw6** [comment]
@mirie 

> Let me know if it's easier for you if I record a Loom of my findings or if you have any questions about my findings so far.

Maybe just to confirm, from what I can tell, there's some question about setting future `published_at`, but it otherwise seems like most of our issues (so far) are "couldn't test" ? (ie, not setup on canary1)

> Also, is it easier to deploy (I see you have this PR approved) and address the remaining bugs as they exist in a smaller PR before we enable the feature flag on Dev.to?

My sense is that we're pretty comfortable that the risks of deploying are minimal -- the things we couldn't test on canary _can_ be tested when deployed to production environments -- and we can roll-back easily if we discover big surprises. (Changes in the PR are code, this isn't a hard-to-roll-back PR, like dropping a column.)

Or: I'm fairly confident that the scope of change we're dealing with here is more likely to have "blown up" spectacularly than to have failed subtly, if we haven't "blown up" in canary, we can go to `main` and roll-back if the subtleties pile up.

**Comment by mirie** [comment]
Okay summarizing some conversations from today in one place:

### Deploy
As this is protected by a feature flag, we are ready to merge and deploy this work and continue QA in the meantime. So far we haven't come across anything major that would hold up a deploy. As we go through the remaining QA. Any fixes can be addressed in a smaller followup PR -- we will discuss in #product-management prioritization of bugs as well. 
^ this means that @lightalloy should be unblocked from the next rendering issue as well.

### Remaining QA

- [ ] @anujbhavsar96 to help with QA on mobile apps (creating, editing, deleting Articles) - iOS and Android. _Note from our conversation: All things look great. I’ve checked today, and they work as expected._
- [ ] @anujbhavsar96 will double check with @rt4914 on any mobile app specific functionality that should be double checked
- [ ] @jaw6 to check on Giphy URL rewrite
- [ ] @jaw6 to check for v0 editor: series functionality works as expected
- [ ] @jaw6 to check if CodePen, CodeSandbox, Instagram, and Medium liquid tags work on Dev.to (both V0 and V1 editors) as they are not working on canary1. 
- [ ] @jaw6 to check Next Tech liquid tag (v0 and v1 editors) on canary1
- [ ] @mirie to check if Spotify, StackBlitz liquid tags work on Dev.to (both V0 and V1 editors) as they are not working on canary1
- [ ] @mirie to check Stackery liquid tag (v0 and v1 editors) on canary1
- [ ] @mirie to check Twitter and Twitter timeline liquid tag after deploy (@jaw6 assuming that we can enable the feature flag for my user only)
- [ ] @mirie to check if <figure> and <figcaption> elements are rewritten/repositioned when the markup is incorrect (v0 and v1 editors)
- [ ] @mirie to check if the following markup is supported on dev.to (Alt-H1, Alt-H2, Table: alignment of items in cell and zebra striping)

### Schedule to Publish
When testing I noticed that there might be a problem with this functionality for the v0 editor. I am going to retest and share my testing steps to confirm that (1) I understand how the feature works and (2) to highlight any issues I find as a result of testing. I'll do that in a **separate issue**

**Comment by mirie** [comment]
In terms of the QA I did (checked items above), everything matched dev.to/looked fine. 

The only thing to note is that I was unsure how to test the Stackery liquid tag. Asked about [here](https://forem-team.slack.com/archives/CAA200SH5/p1670965466022449).

**Comment by jaw6** [comment]
@mirie 

> assuming that we can enable the feature flag for my user only

I mentioned something like this yesterday, but researching this more today, this is something that's _theoretically possible_ with the feature-flag system we have, but requires some changes to the way we use feature-flags in our code. I have updated [the PR](https://github.com/forem/forem/pull/18754) so that this works.

**Comment by jaw6** [comment]
Found & fixed a bug with embeds: https://github.com/forem/forem/pull/18857 (not likely related to the extraction, probably an old bug found through fine-grain QA here)
